contracts/buy/PartyBuy.sol

```diff
@@ -0,0 +1,164 @@
+    uint256 public maxPrice;
+
+    // ============ Events ============
+
+    // emitted when the token is successfully bought
+    event Bought(address triggeredBy, address targetAddress, uint256 ethSpent, uint256 ethFeePaid, uint256 totalContributed);
+
+    // emitted if the Party fails to buy the token before expiresAt
+    // and someone expires the Party so folks can reclaim ETH
+    event Expired(address triggeredBy);
+
+    // ======== Constructor =========
+
+    constructor(
+        address _partyDAOMultisig,
+        address _tokenVaultFactory,
+        address _weth
+    ) Party(_partyDAOMultisig, _tokenVaultFactory, _weth) {}
+
+    // ======== Initializer =========
+
+    function initialize(
+        address _nftContract,
+        uint256 _tokenId,
+        uint256 _maxPrice,
+        uint256 _secondsToTimeout,
+        address _splitRecipient,
+        uint256 _splitBasisPoints,
+        string memory _name,
+        string memory _symbol
+    ) external initializer {
+        // validate maxPrice
+        require(_maxPrice > 0, "PartyBuy::initialize: must set price higher than 0");
+        // initialize & validate shared Party variables
+        __Party_init(_nftContract, _tokenId, _splitRecipient, _splitBasisPoints, _name, _symbol);
+        // set PartyBuy-specific state variables
+        expiresAt = _secondsToTimeout + block.timestamp;
+        maxPrice = _maxPrice;
+    }
+
+    // ======== External: Contribute =========
+
+    /**
+     * @notice Contribute to the Party's treasury
+     * while the Party is still active
+     * @dev Emits a Contributed event upon success; callable by anyone
+     */
+    function contribute() external payable nonReentrant {
+        // require that the new total contributed is not greater than
+        // the maximum amount the Party is willing to spend
+        require(totalContributedToParty + msg.value <= getMaximumContributions(), "PartyBuy::contribute: cannot contribute more than max");
+        // continue with shared _contribute flow
+        _contribute();
+    }
+
+    // ======== External: Buy =========
+
+    /**
+     * @notice Buy the token by calling targetContract with calldata supplying value
+     * @dev Emits a Bought event upon success; reverts otherwise. callable by anyone
+     */
+    function buy(uint256 _value, address _targetContract, bytes calldata _calldata) external nonReentrant {
+        require(
+            partyStatus == PartyStatus.ACTIVE,
+            "PartyBuy::buy: party not active"
+        );
+        // check that value is not zero (else, token will be burned in TokenVault)
+        require(_value > 0, "PartyBuy::buy: can't spend zero");
+        // check that value is not more than the maximum price set at deploy time
+        require(_value <= maxPrice, "PartyBuy::buy: can't spend over max price");
+        // check that value is not more than
+        // the maximum amount the party can spend while paying ETH fee
```

Comment:
should we check the `expiresAt` time?

i.e: should we allow this purchase to go thru even if it's past the expiresAt time?

Reply:
yes, i thought about this and i think we should allow the purchase to go through if it's past the expire time. the purpose of the party is to win the NFT, so if it can do that, that's mission accomplished.
the purpose of expiring is to give people a path to get their money back at some point if it becomes impossible to buy the NFT